### DISCUSSION TITLE
Can pgf support /ShadingType 5?

### DISCUSSION THREAD

**Discussion by cspshare** [pgf-tikz/pgf#1365] [Ideas]
For complex gradients, “/ShadingType 5” is very useful, will pgf have an official implementation for that type of shading in the future? I hope it could support "/ShadingType 5" shading, such as：

```postscript
/inch {72 mul} def
/DeviceRGB setcolorspace
%Define the shading dictionary
<<
    /ShadingType 5 
    /ColorSpace /DeviceRGB 
    /VerticesPerRow 2 
    /DataSource [
        1 inch 1 inch    1 0 1  %Magenta at bottom left 
        7 inch 1 inch    0 1 1  %Cyan at bottom right
        2 inch 10 inch    0 1 1  %Cyan at top left
        8 inch 10 inch    1 0 1  %Magenta at top right
  ]
>> shfill
showpage
```

**Comment by hmenke** [comment]
<details>

<summary>Relevated pages from the <a href="https://opensource.adobe.com/dc-acrobat-sdk-docs/pdfstandards/pdfreference1.7old.pdf">PDF 1.7 Standard</a></summary>

![pdfreference1 7old-319](https://github.com/user-attachments/assets/d808672d-27b1-4961-91cb-652909b9dd1e)
![pdfreference1 7old-320](https://github.com/user-attachments/assets/944e9ba3-b396-4be9-a071-b96b7ad2a522)
![pdfreference1 7old-321](https://github.com/user-attachments/assets/a30c9f0e-6370-44ef-a5eb-a8a4aeb02c89)

</details>

**Comment by hmenke** [comment]
Is there any information on how is the viewer support for this? I don't want to invest into anything that is only implemented by Adobe Reader.

**Reply by cspshare** [reply]
It supported well by kinds of pdf reader and web browsers and with Pretty awesome rendering efficiency. The following tools are what I have tested: Adobe Acrobat, PDF-XChange, SumatraPDF, OKular, Firefox, Google Chrome, Microsoft Edge, Brave.

here is the postscript code:
```postscript
%!PS-Adobe-3.0 EPSF-3.0

<< /PageSize [80 100] >> setpagedevice
/inch {10 mul} def

/DeviceRGB setcolorspace

%Define the shading dictionary
/pic{
<<
    /ShadingType 5 
    /ColorSpace /DeviceRGB 
    /VerticesPerRow 3
    /DataSource [
        0  0    1 0 1  
        40 0    1 0 1  
        80 0    0 1 1 
        0  50    1 0.2 .1 
        40  50    1 0.2 .1 
        80  50    0.2 1 0.1 
        0  100    0 1 1 
        40  100    0 1 1 
        80  100    1 0 1
  ]

>> shfill
}def
pic 
showpage
```

as well as the output pdf:
[ps2test.pdf](https://github.com/user-attachments/files/17495119/ps2test.pdf)

**Comment by jasperhabicht** [comment]
Doesn't PGFplots make use of Type 5 or Type 6 shadings? 

The `hawkdraw` package uses Type 6 shadings to display conic shadings. So, I can say from my personal coding experience that the viewer support seems to be quite good and an implementation is feasible with PDFTeX, LuaTeX and dvipdfmx. 

There will be some endeavour to add shadings to the LaTeX3 kernel, probably also Type 5, 6, 7 shadings (but this might take a while to be realised).
